### ISSUE TITLE
Retrained models really big

### ISSUE THREAD

**Issue by tunguz** [facebook/fb.resnet.torch#40]
Hi,

I have some question about the size of the retrained models. The t7 files I get after retraining are really big - the smallest ones are 1.6 GB for resnet-18. Is there a reason why this is the case? Is there a way to reduce these files? Is it possible to reduce the memory of the files while training them?

Thanks,

Bojan

**Comment by gcr** [comment]
This was discussed in #38, #24 

@colesbury writes: Load the model, run clearState(), and save the model:
```
model = torch.load('model.t7')
model:clearState()
torch.save('model_small.t7', model)

**Comment by tunguz** [comment]
Thanks for your help. So far I have only been using the command-line level training. Where exactly do I have to insert those lines of code?

**Comment by colesbury** [comment]
You can add `model:clearState()` here:

https://github.com/facebook/fb.resnet.torch/blob/master/checkpoints.lua#L29

However, this breaks the sharing `gradInput`, so you should *not* run with `-shareGradInput`. This is why I haven't added it to checkpointing code yet.

**Comment by tunguz** [comment]
Thanks. I guess at this point I care more about `-shareGradInput` than the size of model files, so I think I'll stick with the existing code after all. 

Any idea how much GPU memory is required for retraining 101, 152 and 200 resnets?

**Comment by colesbury** [comment]
So these numbers are for GPU with 12 GB of memory each (e.g. K40, Titan X) with batch size 256. You can use a smaller batch size, but then your on your own when it comes to the LR schedule:

ResNet-101 fits on 4 GPUs and requires ~10.5 GB per GPU
ResNet-152 fits on 8 GPUs and requires ~7 GB per GPU
ResNet-200 fits on 8 GPUS and requires ~9 GB per GPU

(You can also have a larger effective batch size by accumulating the gradients from multiple mini-batches before updating parameters)

**Comment by tunguz** [comment]
Yeah, as I don't see getting 4 - 8 12 GB any time soon, I think I'll stick with lower capacity models for now. :) Thanks for all your great help and insights!

**Comment by pedropgusmao** [comment]
Is it possible, during initializaiton,  to create a dummy model that will only share  with the original training model variables that are useful during testing and then save this dummy model instead? 
For example, inside `main.lua` : 

    -- Create model
    local model, criterion = models.setup(opt, checkpoint)

    -- Create dummy model just for saving 
    local dummymodel = model:clone()
    dummymodel:share(model, 'bias')
    dummymodel:share(model, 'weight')
    dummymodel:share(model, 'running_mean') -- For BatchNorm
    dummymodel:share(model, 'running_var')     -- For BatchNorm
    
    --- Enter for cycle that trains for an epoch ...
    ... 
    -- At the end of for loop epoch   
        checkpoints.save(epoch, dummymodel, trainer.optimState, bestModel)
    end 

Could this help with the `gradInput` sharing?

**Comment by colesbury** [comment]
@pedropgusmao, yeah that might be the best option for now

**Comment by colesbury** [comment]
Fixed in f54d7c806cc52b17978373642fcb7e11e4b4a347
